File: lib/domain/repositories/folder_repository.dart
import '../entities/file_entity.dart';
import '../entities/folder_entity.dart';

class FolderContents {
  final List<FolderEntity> folders;
  final List<FileEntity> files;

  const FolderContents({required this.folders, required this.files});

  int get totalCount => folders.length + files.length;
  bool get isEmpty => folders.isEmpty && files.isEmpty;
}

abstract class FolderRepository {
  /// List root folders.
  Future<List<FolderEntity>> listRootFolders();

  /// List subfolders and files in a folder.
  Future<FolderContents> listFolderContents(String folderId);

  /// Get folder by ID.
  Future<FolderEntity> getFolder(String id);

  /// Create a new folder.
  Future<FolderEntity> createFolder({
    required String name,
    String? parentId,
  });

  /// Rename a folder.
  Future<FolderEntity> renameFolder(String id, String newName);

  /// Move a folder.
  Future<FolderEntity> moveFolder(String id, String? newParentId);

  /// Delete a folder (soft-delete to trash).
  Future<void> deleteFolder(String id);

  /// Download folder as ZIP.
  Future<Stream<List<int>>> downloadFolderZip(String id);
}

File: src/domain/repositories/folder_repository.rs
use anyhow::Result;
use async_trait::async_trait;

use crate::domain::models::folder::Folder;
use crate::domain::models::file::File;

/// Repository interface for folder operations
#[async_trait]
pub trait FolderRepository: Send + Sync {
    /// Get a folder by its ID
    async fn get_folder_by_id(&self, id: &str) -> Result<Option<Folder>>;
    
    /// Get a folder by its path
    async fn get_folder_by_path(&self, path: &str) -> Result<Option<Folder>>;
    
    /// Create a new folder
    async fn create_folder(&self, parent_id: &str, name: &str) -> Result<Folder>;
    
    /// Update folder metadata
    async fn update_folder(&self, folder: &Folder) -> Result<Folder>;
    
    /// Delete a folder
    async fn delete_folder(&self, id: &str) -> Result<()>;
    
    /// Get the root folder
    async fn get_root_folder(&self) -> Result<Folder>;
    
    /// Get all subfolders of a folder
    async fn get_subfolders(&self, folder_id: &str) -> Result<Vec<Folder>>;
    
    /// Get all files in a folder
    async fn get_files_in_folder(&self, folder_id: &str) -> Result<Vec<File>>;
    
    /// Get the contents of a folder (files and subfolders)
    async fn get_folder_contents(&self, folder_id: &str) -> Result<(Vec<File>, Vec<Folder>)>;
    
    /// Move a folder to a different parent
    async fn move_folder(&self, id: &str, new_parent_id: &str) -> Result<Folder>;
    
    /// Rename a folder
    async fn rename_folder(&self, id: &str, new_name: &str) -> Result<Folder>;
    
    /// Set folder as favorite or not
    async fn set_favorite(&self, id: &str, is_favorite: bool) -> Result<Folder>;
}